### PR TITLE
Webcam Demo

### PR DESCRIPTION
**Pull request by ly015**
This PR adds a WebCam demo script with the following features:

1. Read video stream from the webcam
2. Perform object detection and human/**animal** pose estimation
3. Provide better visualization of bbox, object label, and pose
4. Optionally, apply special effects (`sunglasses` or `bug-eye`)
5. Show time-consuming information
6. Optionally, save out the video

### CODE REVIEW

**Comment by innerlee** [comment]
Can it run in realtime?

File: demo/webcam_demo.py [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+                         vis_pose_result)
+
+try:
+    from mmdet.apis import inference_detector, init_detector
+    has_mmdet = True
+except (ImportError, ModuleNotFoundError):
+    has_mmdet = False
+
+
+class StopWatch:
```
Comment (innerlee) [inline demo/webcam_demo.py:18]: Where does this code come from?

**Comment by innerlee** [comment]
There is a difference between webcam demo and video demo. For webcam, the UI should be responsive for slow models and the inference should not take all cpus for small models.
See the webcam in mmaction2 for reference. https://github.com/open-mmlab/mmaction2/blob/master/demo/webcam_demo.py

File: demo/webcam_demo.py (hunk 1 above)
Comment (ly015) [reply demo/webcam_demo.py:18]: It was not adapted from somewhere else.

**Comment by ly015** [comment]
> Can it run in realtime?

Yes almost. Around 20fps if it's not too crowded.

**Comment by innerlee** [comment]
> Around 20fps

Cool. Please check both heavy models and lightweight models. Also monitor the cpu usage.

File: demo/webcam_demo.py (hunk 1 above)
Comment (innerlee) [reply demo/webcam_demo.py:18]: If this util is generally useful, then it can be put into the repo's util

**Comment by ly015** [comment]
> > Around 20fps
> 
> Cool. Please check both heavy models and lightweight models. Also monitor the cpu usage.

I'm afraid it won't work in realtime with heavy detection model. So we use a yolov3 as default. A FasterRCNN with r50 backbone will reduce the fps to 4-5 fps.

File: demo/webcam_demo.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+        return strings
+
+    def reset(self):
+        self._record = defaultdict(list)
+        self._timer_stack = []
+
+
+def apply_bugeye_effect(img,
+                        pose_results,
+                        dataset='TopDownCocoDataset',
```
Comment (jin-s13) [inline demo/webcam_demo.py:109]: I think it is better to use `leye_idx` and `reye_idx` as the input.

Comment (ly015) [reply demo/webcam_demo.py]: fixed

File: demo/webcam_demo.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+            strings.append(f'FPS: {fps:.1f}')
+        strings += [f'{name}: {val:.0f}' for name, val in result.items()]
+        return strings
+
+    def reset(self):
+        self._record = defaultdict(list)
+        self._timer_stack = []
+
+
+def apply_bugeye_effect(img,
```
[the remark this answers is not in the document]
Comment (ly015) [reply demo/webcam_demo.py]: moved them to core/visualization

File: demo/webcam_demo.py (hunk 1 above)
Comment (ly015) [reply demo/webcam_demo.py]: moved to mmpose/utils.py

File: demo/mmdetection_cfg/yolov3_d53_320_273e_coco.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            dict(type='DefaultFormatBundle'),
+            dict(type='Collect', keys=['img'])
+        ])
+]
+data = dict(
+    samples_per_gpu=8,
+    workers_per_gpu=4,
+    train=dict(
+        type=dataset_type,
+        ann_file=data_root + 'annotations/instances_train2017.json',
```
Comment (innerlee) [inline demo/mmdetection_cfg/yolov3_d53_320_273e_coco.py:98]: f-str

File: mmpose/core/visualization/effects.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                            reye_index,
+                            kpt_thr=0.5):
+    """Apply sunglasses effect.
+
+    Args:
+        img (np.ndarray): Image data.
+        pose_results (list[dict]): The pose estimation results containing:
+            - "keypoints" ([K,3]): keypoint detection result in [x, y, score]
+        sunglasses_img (np.ndarray): Sunglasses image with white background.
+        dataset (str): Dataset name (e.g. 'TopDownCocoDataset') to determine
```
Comment (jin-s13) [inline mmpose/core/visualization/effects.py:71]: Wrong docstring

File: mmpose/core/visualization/effects.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
+import cv2
+import numpy as np
+
+
+def apply_bugeye_effect(img,
+                        pose_results,
+                        leye_index,
```
Comment (innerlee) [inline mmpose/core/visualization/effects.py:7]: How about `left_eye_index`?

File: demo/webcam_demo.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+    det_results = []
+    for bbox, label in zip(bboxes, labels):
+        det_result = dict(bbox=bbox, label=label)
+        det_results.append(det_result)
+    return det_results
+
+
+def read_camera():
+    # init video reader
+    print('Thred "input" started')
```
Comment (innerlee) [inline demo/webcam_demo.py:195]: typo

File: demo/webcam_demo.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+        type=int,
+        default=-1,
+        help='Frame buffer size. If set -1, the buffer size will be '
+        'automatically inferred from the display delay time. Deafult: -1')
+
+    parser.add_argument(
+        '--inference_fps',
+        type=int,
+        default=10,
+        help='Maximum inference FPS. This is to limit the resource consumnig '
```
Comment (innerlee) [inline demo/webcam_demo.py:135]: typo

File: demo/webcam_demo.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+        event_inference_done.set()
+
+
+def display():
+    print('Thred "display" started')
+    stop_watch = StopWatch(window=10)
+
+    # initialize result status
+    ts_inference = None  # timestamp of the lastet inference result
+    fps_inference = 0.  # infenrece FPS
```
Comment (innerlee) [inline demo/webcam_demo.py:316]: typo

File: demo/mmdetection_cfg/yolov3_d53_320_273e_coco.py (hunk 4 above)
Comment (ly015) [reply demo/mmdetection_cfg/yolov3_d53_320_273e_coco.py]: done

File: mmpose/core/visualization/effects.py (hunk 5 above)
Comment (ly015) [reply mmpose/core/visualization/effects.py]: fixed

File: mmpose/core/visualization/effects.py (hunk 6 above)
Comment (ly015) [reply mmpose/core/visualization/effects.py]: fixed

File: demo/webcam_demo.py (hunk 7 above)
Comment (ly015) [reply demo/webcam_demo.py]: fixed

File: demo/webcam_demo.py (hunk 8 above)
Comment (ly015) [reply demo/webcam_demo.py]: fixed

File: demo/webcam_demo.py (hunk 9 above)
Comment (ly015) [reply demo/webcam_demo.py:316]: fixed

File: demo/mmdetection_cfg/cascade_rcnn_x101_64x4d_fpn_1class.py [hunk 10]
```diff
@@ -239,17 +239,17 @@
     workers_per_gpu=2,
     train=dict(
         type=dataset_type,
-        ann_file=data_root + 'annotations/instances_train2017.json',
-        img_prefix=data_root + 'train2017/',
+        ann_file=f'{data_root}annotations/instances_train2017.json',
```
Comment (innerlee) [inline demo/mmdetection_cfg/cascade_rcnn_x101_64x4d_fpn_1class.py:242]: use ` f'{data_root}/annotations/instances_train2017.json' ` and remove the trailing ` / ` in data_root

Comment (ly015) [reply demo/mmdetection_cfg/cascade_rcnn_x101_64x4d_fpn_1class.py]: fixed.

**Review by jin-s13 — APPROVED** [review]

**Review by innerlee — APPROVED** [review]

**Pull request merged**